### PR TITLE
[WIP] Causal Input/Output variables, explicit algebraic variables, aliases, and open connections

### PR DESCRIPTION
**Pull request by ChrisRackauckas**

### CODE REVIEW

File: src/systems/abstractsystem.jl [hunk 1]
```diff
@@ -135,12 +137,21 @@ function Base.getproperty(sys::AbstractSystem, name::Symbol)
...
         end
     end
+
+    if :outputs ∈ fieldnames(typeof(sys))
+        i = findfirst(x->convert(Variable,x.lhs).name==name,sys.outputs)
+        if i !== nothing
+            return rename(convert(Variable,sys.outputs[i].lhs),renamespace(sys.name,name))(getfield(sys,:iv)())
+        end
+    end
+
```
Comment (shashi) [inline src/systems/abstractsystem.jl:154]: lol impressive.

**Comment by shashi** [comment]
Ok this is basically ready as a low level infrastructure.

**Comment by ChrisRackauckas** [comment]
🎉 looks good. If it passes let's merge.

**Comment by shashi** [comment]
Some things to review:

- `observed(sys)` will return both the observed of `sys` and the namespaced obsreved of `sys.systems` in a vector of equations
- `equations(sys)` will try to do the simple `lhs = rhs` substitution of the observed variables in all the equations of `sys` and `sys.systems` (you can disable this with `remove_aliases=false` kwarg)

File: src/systems/reaction/reactionsystem.jl [hunk 2]
```diff
@@ -294,6 +299,7 @@ explicitly on the independent variable (usually time).
 function ismassaction(rx, rs; rxvars = get_variables(rx.rate),
                               haveivdep = any(var -> isequal(rs.iv,convert(Variable,var)), rxvars),
                               stateset = Set(states(rs)))
+    return !(haveivdep || rx.only_use_rate || any(convert(Variable,rxv) in states(rs) for rxv in rxvars))
```
Comment (isaacsas) [inline src/systems/reaction/reactionsystem.jl:302]: @ChrisRackauckas Any reason you cut off the code path here? This was the old version of `ismassaction`, which was having issues when profiling.

**Pull request merged**